### PR TITLE
revamping single visit catalog reader

### PR DESCRIPTION
**Pull request by yymao**
@fjaviersanchez when I was reviewing your [PR](https://github.com/LSSTDESC/gcr-catalogs/pull/247), I found that there are too much repeated effort (converting butler table to dict, concatenate datasets, etc). 

I hence took the liberty of revamping the reader to basically a simple wrapper around butler, and reuse as much as the native features of butler and GCR as possible. 

One thing that I am not sure is the `calib.getMagnitude` conversion. For now I use the GCR's quantity modifiers to implement it. However, if the zero point if different detector by detector then this would not be correct. Do you know if the conversion needs to be done differently detector by detector? 

Otherwise, can you give this a test and see if it satisfies all your needs? Thanks.

### CODE REVIEW

**Comment by fjaviersanchez** [comment]
Thanks a lot @yymao . The zeropoint actually changes detector by detector (the variation is small but exists). There's a requirement on the maximum variation on the zeropoints (3.3.4.2 in the [LSST-SRD](https://docushare.lsst.org/docushare/dsweb/Get/LPM-17)) so we could play with that but I'd rather keep the zeropoints somehow. I'll give it a try :)

File: GCRCatalogs/butler_interface.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    def _subclass_init(self, repo, filter_band, visit, detector=None, **kwargs):
+
+        dataId = {'filter': filter_band, 'visit': int(visit)}
+        if detector is not None:
+            dataId['detector'] = int(detector)
+
+        super()._subclass_init(repo, 'src', dataId, **kwargs)
+
+        self._calib = None
```
Comment (fjaviersanchez) [inline GCRCatalogs/butler_interface.py]: You can either create a list of `Calib` objects (which may be useful for checking the spatial homogeneity of the zeropoints for example) or leave it as it is.

File: GCRCatalogs/butler_interface.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        for dataId in self._dataId_cache:
+            calexp_md = self._get_data(dataId, 'calexp_md')
+            if calexp_md is not None:
+                self._calib = Calib(calexp_md)
+                break
+
+        if self._calib is not None:
```
Comment (fjaviersanchez) [inline GCRCatalogs/butler_interface.py]: if you create the list of calibs, together with a list of lengths of each catalog, I guess that we would be able to still use the quantity modifier somehow to add the magnitudes?

If not, we could just do the quantity modification for each sensor/visit.

**Comment by yymao** [comment]
@fjaviersanchez thanks for the quick feedback. In this case we can't use quantity modifiers, but we can do the magnitude calibration at the quantity getter step. So I made this new [method](https://github.com/fjaviersanchez/gcr-catalogs/pull/2/files#diff-eb91a4dda2f7bd3e52378d28ec5e4945R100). 

See if this works for you?

**Comment by fjaviersanchez** [comment]
Thanks a lot! It works. Merged!

**Pull request merged**